runtime/ioext/asyncpipe.go

```diff
@@ -0,0 +1,128 @@
+package ioext
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+// ErrPipeFull is returned from AsyncPipeWriter.Write if the pipes capacity has
+// been reached.
+var ErrPipeFull = errors.New("The internal pipe buffer have reached its capacity")
+
+// AsyncPipeReader is a reading side of an AsyncPipe, similar to io.PipeReader.
+type AsyncPipeReader struct {
+	*AsyncPipeWriter
+}
+
+// AsyncPipeWriter is a writing side of a BlockedPipe, similar to io.PipeWriter.
+type AsyncPipeWriter struct {
+	c        sync.Cond
+	m        sync.Mutex
+	buffer   []byte
+	closed   bool
+	broken   bool
+	tell     chan<- int
+	capacity int
+}
+
+// AsyncPipe is similar to io.Pipe() except that writes isn't blocking, instead
+// data will be added to an internal buffer that can grow up to specified
+// capacity.
+// Additionally, you may supply a channel tell that will be told whenever
+// N bytes have been read, so that more bytes can be requested to be written.
+//
+// This pipe kind is useful when implementing congestion control.
```

Comment:
no, it can't the semantics for sending when buffer is full is non-blocking, returns an error..
Also channels would be crazy slow... a channel has a cost per message, if each byte is a message.. well, let's just leave at that would be very very slow, unless go has some builtin magic for chan byte, which I doubt...